pipeline/runtime/runtestsV2.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	outputFile := fmt.Sprintf("%s/%s-output.env", pipeline.SharedVolPath, step.ID)
+	step.Envs["DRONE_OUTPUT"] = outputFile
+
+	if len(r.Outputs) > 0 {
+		step.Command[0] += getOutputsCmd(step.Entrypoint, r.Outputs, outputFile)
+	} else if len(r.OutputVars) > 0 {
+		step.Command[0] += getOutputVarCmd(step.Entrypoint, r.OutputVars, outputFile)
+	}
+
+	artifactFile := fmt.Sprintf("%s/%s-artifact", pipeline.SharedVolPath, step.ID)
+	step.Envs["PLUGIN_ARTIFACT_FILE"] = artifactFile
+
+	if metadataFile, found := step.Envs["PLUGIN_METADATA_FILE"]; found {
+		step.Envs["PLUGIN_METADATA_FILE"] = fmt.Sprintf("%s/%s-%s", pipeline.SharedVolPath, step.ID, metadataFile)
+	}
+
+	exited, err := engine.Run(ctx, step, out, r.LogDrone)
+	collectionErr := collectTestReportsAndCg(ctx, log, r, start, step.Name, tiConfig)
+	if err == nil {
+		err = collectionErr
+	}
+
+	exportEnvs, _ := fetchExportedVarsFromEnvFile(exportEnvFile, out)
+	artifact, _ := fetchArtifactDataFromArtifactFile(artifactFile, out)
+
+	if exited != nil && exited.Exited && exited.ExitCode == 0 {
+		outputs, err := fetchExportedVarsFromEnvFile(outputFile, out) //nolint:govet
+		if len(r.Outputs) > 0 {
+			outputsV2 := []*api.OutputV2{}
+			for _, output := range r.Outputs {
+				if _, ok := outputs[output.Key]; ok {
+					outputsV2 = append(outputsV2, &api.OutputV2{
+						Key:   output.Key,
+						Value: outputs[output.Key],
+						Type:  output.Type,
+					})
+				}
+			}
+			return exited, outputs, exportEnvs, artifact, outputsV2, "", err
+		} else if len(r.OutputVars) > 0 {
+			// only return err when output vars are expected
+			return exited, outputs, exportEnvs, artifact, nil, "", err
+		}
+		return exited, outputs, exportEnvs, artifact, nil, "", nil
+	}
+	return exited, nil, exportEnvs, artifact, nil, "", err
+}
+
+func collectTestReportsAndCg(ctx context.Context, log *logrus.Logger, r *api.StartStepRequest, start time.Time, stepName string, tiConfig *tiCfg.Cfg) error {
+	cgStart := time.Now()
+
+	cgErr := callgraph.Upload(ctx, stepName, time.Since(start).Milliseconds(), log, cgStart, tiConfig, outDir)
+	if cgErr != nil {
+		log.WithField("error", cgErr).Errorln(fmt.Sprintf("Unable to collect callgraph. Time taken: %s", time.Since(cgStart)))
+		cgErr = fmt.Errorf("failed to collect callgraph: %s", cgErr)
+	}
+
+	reportStart := time.Now()
+	crErr := report.ParseAndUploadTests(ctx, r.TestReport, r.WorkingDir, stepName, log, reportStart, tiConfig, r.Envs)
+	if crErr != nil {
+		log.WithField("error", crErr).Errorln(fmt.Sprintf("Failed to upload report. Time taken: %s", time.Since(reportStart)))
+	}
+	return cgErr
+}
+
+// Second parameter in return type (bool) is will be used to decide whether the filter file should be created or not.
+// In case of running all the cases no filter file should be created.
+func getTestsSelection(ctx context.Context, fs filesystem.FileSystem, stepID, workspace string, log *logrus.Logger,
```

Comment:
Can we reuse this method from the existing runtest step?

Reply:
We could have but here some things have changed not eveything which is being used for runtest are used for runtestV2 because runner interface is being used for some calls there and parameters like "language", "build tool" etc are required in order to create runner. So in getTestsSelection some methods are being called by runner and also for **checkForBazelOptimization** it requires those params like "language", "build tool" . Even though many things are same there are still some changes. That's why thought of creating this method for V2